### ISSUE TITLE
Mpi/crocoddyl topic/action unittests - [merged]

### ISSUE THREAD

**Issue by wxmerkt** [loco-3d/crocoddyl#513]
In GitLab by @MaximilienNaveau on Jul 8, 2019, 18:33

_Merges mpi/crocoddyl-topic/action_unittests -> cpp_devel_

This merge request implements the numerical differentiation of order 1 of the action and its unittests

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 12:38

This is a private function and it should be also rename to `assertStableStateFD`.

Additionally we should use `const Eigen::Ref`, i.e.
```cpp
void assertStableStateFD(const Eigen::Ref<const Eigen::VectorXd>& x);
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 12:40

Change to
```cpp
#ifndef CROCODDYL_CORE_ACTIONS_ACTION_NUM_DIFF_HPP_
#define CROCODDYL_CORE_ACTIONS_ACTION_NUM_DIFF_HPP_
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 12:45

Change to:
```cpp
ActionDataNumDiff(Model* const model_num_diff)
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 12:51

To make things less verbose I suggest to get the different dimensions only once, i.e.:
```cpp
const int& ndx = model->get_ndx();
const int& nu = model->get_nu();
const int& ncost = model->get_ncost();
.
.
// data_x
for (unsigned i = 0; i < ndx; ++i) {
  data_x.push_back(model_num_diff->get_model().createData());
}
.
.
.
// Rx
Rx.resize(ncost, ndx);
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 12:53

I would like to always include header files like this:
```cppp
#include <crocoddyl/core/actions/action-num-diff.hpp>
```
instead of
```cpp
#include "crocoddyl/core/actions/action-num-diff.hpp"
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 12:54

if it's not needed, then remove it :)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:00

Do we need the template here by the way?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:01

Why this?

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:01

I am not sure it is a good idea, because this is "" correspond to a "local include, while <> means that the header is somewhere in another package. I am correct?

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:02

My bad, I will correct it.

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:03

You need to cast the pointer inside the shared_ptr to the good type: `ActionDataNumDiff` instaed of `ActionDataAbstract`.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:05

You could name this as d, and use data instead of data_num_diff.

It'll make easier to debug and read the code :)

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:07

Please add the TODO keyword, and an issue in gepgitlab.

I agree with you that we should later come back to this topic.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:08

Same TODO keyword

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:08

Why do you include the assert here?

I know that the function is empty for the moment.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:10

There is an error here. You should write:
```cpp
data_num_diff->Lxu = data_num_diff->Rx.transpose() * data_num_diff->Ru;

```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:11

are you planning to tackle it on another PR?

If so, for me it's OK. Just add the TODO keyword and an issue in getgilab.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:18

I know that we have a problem with the CI :(. Did you try locally the unit-test? Are they passing (all of c++ and python bindings ones)?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:23

I'm OK if we want to push with this standard.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 13:26

Yes, I think so

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:38

I already added a ticket in the gitlab board: https://gepgitlab.laas.fr/loco-3d/crocoddyl/issues/181

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:38

is this TODO or @todo?

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:39

because it was there in the python code...

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:40

Oo! well seen! Thanks

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:41

I will add the todo, I created this issue https://gepgitlab.laas.fr/loco-3d/crocoddyl/issues/181

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:45

Well, a simple ModelAbstract pointer should be enough no?

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 10, 2019, 13:49

C++ tests passes I make sure of that. For the bindings I am not sure... I can have an addiditional look. This week.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 10, 2019, 16:36

It's weird, there is something wrong.

I just pushed a code and it passes 16.04 for me (see https://gepgitlab.laas.fr/cmastall/crocoddyl/pipelines/4898)

@nim65s could you explain me why the other jobs aren't passing? It's important to have them working properly. With the c++ development, many things could wrong easily.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 11, 2019, 08:41

I suggest to do it like this:
```cpp
// TODO: @Carlos This is my TODO task
```

What do you think?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 11, 2019, 08:44

Merci!

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 11, 2019, 08:47

When you run `calc` (or `calcDiff`), in an action model, you might need to access a membership function of the derived action model class.

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:52

Well, I prefer to avoid 1 letter variable when its not counters. I think the difference between data and data_num_diff is quiet clear. It also easily indicates the type here.

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:59

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/172/diffs?diff_id=960&start_sha=f00d2d22df63658894ffce4419b4696fe904f13e#43982762eadec6f640db2d3356c26f67158de3cd_39_27)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:59

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/172/diffs?diff_id=960&start_sha=f00d2d22df63658894ffce4419b4696fe904f13e#43982762eadec6f640db2d3356c26f67158de3cd_9_9)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:59

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/172/diffs?diff_id=960&start_sha=f00d2d22df63658894ffce4419b4696fe904f13e#43982762eadec6f640db2d3356c26f67158de3cd_102_103)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:59

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/172/diffs?diff_id=960&start_sha=f00d2d22df63658894ffce4419b4696fe904f13e#dfeda92a6d447ee361732bd849c15ed0ae58b2b6_10_10)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:59

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/172/diffs?diff_id=960&start_sha=f00d2d22df63658894ffce4419b4696fe904f13e#dfeda92a6d447ee361732bd849c15ed0ae58b2b6_94_93)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 11, 2019, 16:59

added 1 commit

<ul><li>3002a74e - [review] update the code accoding to the reviews</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/172/diffs?diff_id=960&start_sha=f00d2d22df63658894ffce4419b4696fe904f13e)

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 12, 2019, 08:10

What about to rename data_num_diff to data_nd?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 12, 2019, 08:15

Sorry, I didn't refer to the shared_ptr. My question is regarding `recalc`. But I guess it's not bad idea to do it like this. Let's keep it as you did :)

**Comment by wxmerkt** [comment]
In GitLab by @MaximilienNaveau on Jul 12, 2019, 08:46

I merely copied the python code here.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 14, 2019, 17:07

I meant why you don't include it

I resolved this discussion since you're going to tackle later

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 14, 2019, 17:12

The CI is failing due to renaming of files. I will handle myself

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 14, 2019, 17:13

resolved all discussions

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 14, 2019, 17:14

@MaximilienNaveau The renamed of files produces a problem in the CI. Please be aware that changes in the core library needs to be propagated to the bindings too.

For this time, I will handle this issue. I think it's more efficient for both.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 14, 2019, 17:14

mentioned in commit d03decc842aa9a6d1710e56354427042877d275e

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Jul 14, 2019, 17:14

merged
